File: packages/nonoka-opencode-provider/src/protocol.ts
/**
 * NDJSON protocol messages exchanged between the OpenCode provider and
 * the nonoka-cli --server backend.
 */

export const NONOKA_MESSAGE_ROLES = {
  system: 'system',
  user: 'user',
  assistant: 'assistant',
  tool: 'tool',
} as const;

export type NonokaMessageRole = keyof typeof NONOKA_MESSAGE_ROLES;

export const NONOKA_BRIDGE_PROTOCOL_VERSION = '1.1';
export const NONOKA_PROVIDER_VERSION = '0.2.19';
export const NONOKA_REQUIRED_CAPABILITIES = [
  'external_tool_receipts',
  'persistent_runtime_limits',
  'termination_reasons',
  'tool_approval_resume',
  'typed_verification_receipts',
] as const;

export interface NonokaProtocolContract {
  version: string;
  required_capabilities: string[];
  provider_version?: string;
}

export interface NonokaChatToolCall {
  id: string;
  name: string;
  arguments: string;
}

export type ObservationCompleteness = 'complete' | 'partial' | 'unknown';

export interface NonokaExternalToolReceipt {
  result: unknown;
  exit_code?: number;
  host?: string;
  artifact_ref?: string;
  output_kind?: string;
  original_bytes?: number;
  truncated?: boolean;
  completeness: ObservationCompleteness;
  workspace?: Record<string, unknown>;
  effect?: Record<string, unknown>;
  verification?: NonokaVerificationReceipt;
}

export type NonokaVerificationStatus = 'passed' | 'failed' | 'unavailable' | 'not_run';
export type NonokaVerificationLevel = 'focused' | 'full';
export type NonokaVerificationKind = 'test' | 'build' | 'lint' | 'typecheck' | 'custom';

export interface NonokaVerificationReceipt {
  status: NonokaVerificationStatus;
  level: NonokaVerificationLevel;
  kind: NonokaVerificationKind;
  command: string;
  cwd: string;
  exit_code?: number;
  timed_out: boolean;
  timeout_seconds?: number;
  truncated: boolean;
  completeness: ObservationCompleteness;
  collected_tests?: number;
  executed_tests?: number;
  deselected_tests?: number;
  summary?: string;
  failure_summary?: string;
  artifact_ref?: string;
}

export interface NonokaChatMessage {
  role: NonokaMessageRole;
  content: string;
  tool_call_id?: string;
  tool_calls?: NonokaChatToolCall[];
  result?: unknown;
}

export interface ExternalToolDefinition {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
}

export interface ExternalMCPToolDefinition {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
}

export interface ExternalMCPServerDefinition {
  name: string;
  description?: string;
  tools: ExternalMCPToolDefinition[];
}

export interface ExternalSkillToolDefinition {
  name: string;
  description: string;
  parameters: Record<string, unknown>;
}

export interface ExternalSkillDefinition {
  name: string;
  description?: string;
  tools: ExternalSkillToolDefinition[];
  system_prompt?: string;
  activation_prompt?: string;
}

export const NONOKA_INBOUND_TYPES = {
  chat: 'chat',
  approval: 'approval',
  cancel: 'cancel',
} as const;

export type NonokaInboundEventType = keyof typeof NONOKA_INBOUND_TYPES;

export interface NonokaChatRequest {
  type: typeof NONOKA_INBOUND_TYPES.chat;
  protocol?: NonokaProtocolContract;
  purpose?: 'chat' | 'title';
  messages: NonokaChatMessage[];
  tools?: ExternalToolDefinition[];
  external_mcp_servers?: ExternalMCPServerDefinition[];
  external_skills?: ExternalSkillDefinition[];
  session_id?: string;
  new_session?: boolean;
  cwd: string;
  model?: string;
  temperature?: number;
  max_turns?: number;
  timeout_seconds?: number;
  wall_timeout_seconds?: number;
  tool_budget?: number;
  max_context_bytes?: number;
  max_external_result_bytes?: number;
  require_workspace_mutation?: boolean;
  require_observed_effect?: boolean;
  require_focused_verification?: boolean;
  verification_enforcement?: 'strict' | 'advisory';
  max_completion_corrections?: number;
  request_id?: string;
}

export interface NonokaApprovalMessage {
  type: typeof NONOKA_INBOUND_TYPES.approval;
  id: string;
  approved: boolean;
  modified_args?: Record<string, unknown>;
}

export interface NonokaCancelMessage {
  type: typeof NONOKA_INBOUND_TYPES.cancel;
  request_id?: string;
}

export type NonokaInboundMessage = NonokaChatRequest | NonokaApprovalMessage | NonokaCancelMessage;

export const NONOKA_OUTBOUND_TYPES = {
  protocol_ack: 'protocol_ack',
  session_init: 'session_init',
  text_delta: 'text_delta',
  tool_call: 'tool_call',
  tool_result: 'tool_result',
  tool_call_progress: 'tool_call_progress',
  approval_request: 'approval_request',
  debug: 'debug',
  finish: 'finish',
  error: 'error',
} as const;

export type NonokaOutboundEventType = keyof typeof NONOKA_OUTBOUND_TYPES;

export interface NonokaProtocolAckEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.protocol_ack;
  version: string;
  capabilities: string[];
  cli_version: string;
  framework_version: string;
}

export interface NonokaSessionInitEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.session_init;
  session_id: string;
}

export interface NonokaTextDeltaEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.text_delta;
  text: string;
}

export interface NonokaToolCallEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.tool_call;
  tool_call_id: string;
  tool_name: string;
  args?: unknown;
  metadata?: Record<string, unknown>;
}

export interface NonokaToolResultEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.tool_result;
  tool_call_id: string;
  tool_name: string;
  content: string;
  result?: unknown;
  is_error?: boolean;
}

export interface NonokaToolCallProgressEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.tool_call_progress;
  tool_call_index: number;
  tool_name?: string;
  argument_chars: number;
}

export interface NonokaApprovalRequestEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.approval_request;
  id: string;
  tool_call_id: string;
  tool_name: string;
  args?: unknown;
}

export interface NonokaDebugEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.debug;
  level: 'info' | 'warning' | 'error';
  message: string;
  payload?: Record<string, unknown>;
}

export const NONOKA_FINISH_REASONS = {
  stop: 'stop',
  error: 'error',
  cancel: 'cancel',
  approval_required: 'approval_required',
  tool_calls: 'tool_calls',
} as const;

export type NonokaFinishReason = keyof typeof NONOKA_FINISH_REASONS;

export interface NonokaFinishEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.finish;
  finish_reason: NonokaFinishReason;
  termination?: Record<string, unknown>;
  runtime?: Record<string, unknown>;
}

export interface NonokaErrorEvent {
  type: typeof NONOKA_OUTBOUND_TYPES.error;
  message: string;
  code?: string;
  retryable?: boolean;
  details?: Record<string, unknown>;
}

export type NonokaOutboundEvent =
  | NonokaProtocolAckEvent
  | NonokaSessionInitEvent
  | NonokaTextDeltaEvent
  | NonokaToolCallEvent
  | NonokaToolResultEvent
  | NonokaToolCallProgressEvent
  | NonokaApprovalRequestEvent
  | NonokaDebugEvent
  | NonokaFinishEvent
  | NonokaErrorEvent;

export function parseOutboundLine(line: string): NonokaOutboundEvent | null {
  const trimmed = line.trim();
  if (!trimmed) return null;
  return JSON.parse(trimmed) as NonokaOutboundEvent;
}

export function encodeChatRequest(req: NonokaChatRequest): string {
  return JSON.stringify(req);
}

export function encodeApprovalMessage(req: NonokaApprovalMessage): string {
  return JSON.stringify(req);
}

export function encodeCancelMessage(req: NonokaCancelMessage): string {
  return JSON.stringify(req);
}

File: src/nonoka_cli/bridge/protocol.py
"""NDJSON message protocol for the nonoka-cli server mode.

All messages on stdin/stdout are single-line JSON objects separated by ``\n``.
This protocol is intentionally minimal for Phase 1 and targets the
Vercel AI SDK custom provider bridge.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

# --------------------------------------------------------------------------- #
# Inbound messages: OpenCode provider -> nonoka-cli
# --------------------------------------------------------------------------- #


class ToolCall(BaseModel):
  """A tool call attached to an assistant message."""

  id: str
  name: str
  arguments: str


class ChatMessage(BaseModel):
  """A single message in the chat history."""

  role: Literal["system", "user", "assistant", "tool"]
  content: str
  tool_call_id: str | None = None
  tool_calls: list[ToolCall] | None = None
  result: Any = None


class ExternalToolDefinition(BaseModel):
  """A tool definition supplied by an external host (e.g. OpenCode)."""

  name: str
  description: str
  parameters: dict[str, Any]


class ExternalMCPToolDefinition(BaseModel):
  """A tool provided by an external host-managed MCP server."""

  name: str
  description: str
  parameters: dict[str, Any]


class ExternalMCPServerDefinition(BaseModel):
  """An external host-managed MCP server definition."""

  name: str
  description: str = ""
  tools: list[ExternalMCPToolDefinition]


class ExternalSkillToolDefinition(BaseModel):
  """A tool provided by an external host-managed skill."""

  name: str
  description: str
  parameters: dict[str, Any]


class ExternalSkillDefinition(BaseModel):
  """An external host-managed skill definition."""

  name: str
  description: str = ""
  tools: list[ExternalSkillToolDefinition]
  system_prompt: str = ""
  activation_prompt: str = ""


class ChatRequest(BaseModel):
  """Request nonoka-cli to run one user turn."""

  type: Literal["chat"] = "chat"
  purpose: Literal["chat", "title"] = "chat"
  messages: list[ChatMessage]
  tools: list[ExternalToolDefinition] | None = None
  external_mcp_servers: list[ExternalMCPServerDefinition] | None = None
  external_skills: list[ExternalSkillDefinition] | None = None
  session_id: str | None = None
  new_session: bool = False
  cwd: str = Field(default=".")
  model: str | None = None
  temperature: float | None = None
  max_turns: int | None = Field(default=None, ge=1)
  timeout_seconds: float | None = Field(default=None, gt=0)
  wall_timeout_seconds: float | None = Field(default=None, gt=0)
  tool_budget: int | None = Field(default=None, ge=1)
  max_context_bytes: int | None = Field(default=None, ge=1)
  max_external_result_bytes: int | None = Field(default=None, ge=1)
  require_workspace_mutation: bool = False
  require_observed_effect: bool = False
  request_id: str | None = None


class CancelRequest(BaseModel):
  """Request cancellation of the bridge's currently active chat turn."""

  type: Literal["cancel"] = "cancel"
  request_id: str | None = None


InboundMessage = ChatRequest | CancelRequest


# --------------------------------------------------------------------------- #
# Outbound messages: nonoka-cli -> OpenCode provider
# --------------------------------------------------------------------------- #


class SessionInitEvent(BaseModel):
  """Emitted on the first response so the provider can persist the session id."""

  type: Literal["session_init"] = "session_init"
  session_id: str


class TextDeltaEvent(BaseModel):
  """Incremental assistant text."""

  type: Literal["text_delta"] = "text_delta"
  text: str


class FinishEvent(BaseModel):
  """A single assistant turn finished."""

  type: Literal["finish"] = "finish"
  finish_reason: Literal["stop", "error", "cancel", "approval_required", "tool_calls"]
  termination: dict[str, Any] | None = None
  runtime: dict[str, Any] | None = None


class ToolCallEvent(BaseModel):
  """A tool call was requested by the agent (observation only)."""

  type: Literal["tool_call"] = "tool_call"
  tool_call_id: str
  tool_name: str
  args: Any = None
  metadata: dict[str, Any] | None = None


class ToolResultEvent(BaseModel):
  """A tool call finished (observation only)."""

  type: Literal["tool_result"] = "tool_result"
  tool_call_id: str
  tool_name: str
  content: str
  result: Any = None
  is_error: bool = False


class ApprovalRequestEvent(BaseModel):
  """A tool operation requires human approval."""

  type: Literal["approval_request"] = "approval_request"
  id: str
  tool_call_id: str
  tool_name: str
  args: Any = None


class DebugEvent(BaseModel):
  """Structured debug information for adapter/frontend logs.

  OpenCode does not render this event; it is emitted only when
  ``NONOKA_DEBUG=1`` so that developers can inspect message order, tool
  lists, and session state without polluting the TUI.
  """

  type: Literal["debug"] = "debug"
  level: Literal["info", "warning", "error"] = "info"
  message: str
  payload: dict[str, Any] | None = None


class ErrorEvent(BaseModel):
  """Fatal or terminal error."""

  type: Literal["error"] = "error"
  message: str
  code: str | None = None
  retryable: bool | None = None
  details: dict[str, Any] | None = None


OutboundMessage = (
  SessionInitEvent
  | TextDeltaEvent
  | ToolCallEvent
  | ToolResultEvent
  | ApprovalRequestEvent
  | FinishEvent
  | DebugEvent
  | ErrorEvent
)


# --------------------------------------------------------------------------- #
# Helpers
# --------------------------------------------------------------------------- #


def parse_inbound_line(line: str) -> InboundMessage | None:
  """Parse a single inbound NDJSON line into a typed message.

  Args:
    line: Raw JSON line from stdin.

  Returns:
    A typed message, or None if the line is empty/whitespace.

  Raises:
    ValueError: If the JSON is invalid or the type is unknown.
  """
  line = line.strip()
  if not line:
    return None

  data: dict[str, Any] = __import__("json").loads(line)
  msg_type = data.get("type")

  if msg_type == "chat":
    return ChatRequest.model_validate(data)
  if msg_type == "cancel":
    return CancelRequest.model_validate(data)

  raise ValueError(f"Unknown inbound message type: {msg_type}")


def encode_outbound_message(msg: OutboundMessage) -> str:
  """Serialize an outbound message to a single NDJSON line."""
  return msg.model_dump_json(exclude_none=True)
